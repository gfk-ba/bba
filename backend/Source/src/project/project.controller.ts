/**
 * This files contains the ProjectController
 */

/**
 * import all the stuff we need from NestJS and our code
 */
import { AuthGuard } from '@nestjs/passport';
import {
  Body,
  Controller,
  Get,
  HttpException,
  HttpStatus,
  Post,
  Req,
  UseGuards,
} from '@nestjs/common';
import { Project } from './project.entity';
import { ProjectService } from './project.service';
import { UserDataService } from '../shared/auth/user-data.service';

/**
 * The ProjectController provides methods to get all projects [[findAll]] for a user and to create a new project [[create]]
 * All methods are accessible at http://SERVER:PORT/projects/...
 */
@Controller('projects')
export class ProjectController {
  /**
   * Use dependency injection to get access to the [[ProjectService]] and the [[UserDataService]].
   * @param projectService
   * @param userDataService
   */
  constructor(
    private readonly projectService: ProjectService,
    private readonly userDataService: UserDataService,
  ) {}

  /**
   * This async function delivers the list of all projects for the given user
   * This function is accessible by a HTTP GET call to http://SERVER:PORT/projects
   * This route is restricted by [[AuthGuard]]
   * @param requestData We need the request data to get the information about the user for whom we retrieve the data.
   * @returns The list of all projects for the given user.
   */
  @Get()
  @UseGuards(AuthGuard())
  async findAll(@Req() requestData): Promise<Project[]> {
    const userId = this.userDataService.getUserData(requestData);
    if (!userId) {
      throw new HttpException(
        'userId missing in ProjectController.findAll',
        HttpStatus.EXPECTATION_FAILED,
      );
    }

    return await this.projectService.findAll(userId);
  }

  /**
   * This async function stores a new project in the database
   * This function is accessible by a HTTP POST call to http://SERVER:PORT/projects
   * This route is restricted by [[AuthGuard]]
   * @param projectDto The project data (name and description) as a ProjectDto
   * @param requestData We need the request data to get the information about the user for whom we store the project.
   * @return das in der Datenbank erstellte Projekt
   */
  @Post()
  @UseGuards(AuthGuard())
  async create(
    @Body() projectDto: ProjectDto,
    @Req() requestData,
  ): Promise<Project> {
    const userId = this.userDataService.getUserData(requestData);
    return await this.projectService.create(userId, projectDto);
  }
}
