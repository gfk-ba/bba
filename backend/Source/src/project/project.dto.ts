/**
 * The DTO (data transfer object) used to transport the project data from the frontend to the backend.
 */
interface ProjectDto {
  /**
   * The name of the project.
   */
  readonly name: string;
  /**
   * The description of the project.
   */
  readonly description: string;
}
