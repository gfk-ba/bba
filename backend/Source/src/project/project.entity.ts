/**
 * This files contains the entity class for the document
 */

/**
 * import all decorator from typeorm
 */
import { Column, Entity, PrimaryGeneratedColumn } from 'typeorm';

/**
 * The class Project is used with the ORM TypeORM to access data in the database table projects.
 */
@Entity('projects')
export class Project {
  /**
   * The ID (primary key) of the table (autogenerated)
   */
  @PrimaryGeneratedColumn()
  id: number;

  /**
   * The name of the project
   */
  @Column({ type: 'varchar', length: 100 })
  name: string;

  /**
   * The id of the user who owns the project.
   * Because of SQL problems (userId can only be used in ""), we use user_id as database column.
   */
  @Column({ type: 'int', name: 'user_id' })
  userId: number;

  /**
   * The description of the project.
   */
  @Column({ type: 'varchar', length: 255 })
  description: string;
}
