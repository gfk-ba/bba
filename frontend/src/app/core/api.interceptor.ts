import { Injectable } from '@angular/core';
import { Reset } from '@dcs/ngx-tools';
import { Store } from '@ngrx/store';
import { Observable } from 'rxjs';
import { catchError, flatMap, map, take } from 'rxjs/operators';
import { State } from '../reducers';
import { tokenSelector } from '../reducers/login/login.selectors';
import {
  HttpEvent,
  HttpHandler,
  HttpInterceptor,
  HttpRequest,
  HttpErrorResponse,
} from '@angular/common/http';

@Injectable()
export class ApiInterceptor implements HttpInterceptor {
  constructor(private store: Store<State>) {}

  intercept(request: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
    return this.store.pipe(
      take(1),
      map(state => tokenSelector(state)),
      flatMap(token => {
        request = request.clone({
          setHeaders: {
            Authorization: `Bearer ${token.accessToken}`,
          },
        });
        return next.handle(request).pipe(
          catchError((e: HttpErrorResponse) => {
            if (e.status === 401) {
              this.store.dispatch(new Reset());
            }
            throw e;
          })
        );
      })
    );
  }
}
