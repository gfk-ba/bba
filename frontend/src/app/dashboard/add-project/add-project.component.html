<form [formGroup]="form" class="form">
  <div class="form-group">
    <label for="add-project-name-input">Title (must be unique)</label>
    <input type="text" id="add-project-name-input" placeholder="Title" class="form-control" formControlName="name">
  </div>

  <div class="form-group">
    <label for="add-project-description-input">Description</label>
    <textarea id="add-project-description-input" rows="10" placeholder="Description" class="form-control"
      formControlName="description"></textarea>
  </div>
</form>

<footer>
  <button type="button" class="btn btn-danger" (click)="cancel.next()">cancel</button>
  <button type="button" class="btn btn-primary" [disabled]="!form.valid" (click)="submit()">create</button>
</footer>
