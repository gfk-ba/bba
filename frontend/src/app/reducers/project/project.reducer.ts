import { ActionReducer, combineReducers } from '@ngrx/store';

import { ICurrentProjectState } from './current-project/current-project.manager';
import { currentProject } from './current-project/current-project.reducer';
import { IProjectsListState } from './projects-list/projects-list.manager';
import { projectsList } from './projects-list/projects-list.reducer';

export interface State {
  currentProject: ICurrentProjectState;
  projectsList: IProjectsListState;
}

export const reducer: ActionReducer<State> = combineReducers({
  currentProject,
  projectsList,
});
