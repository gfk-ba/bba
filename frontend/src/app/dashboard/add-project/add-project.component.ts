import { AbstractControl, FormBuilder, FormGroup, Validators } from '@angular/forms';
import { Project } from '../../reducers/project/models/project.class';
import {
  Component,
  EventEmitter,
  Input,
  OnChanges,
  Output,
  SimpleChanges,
  ChangeDetectionStrategy,
} from '@angular/core';

export function uniqueProjectValidator(projects: Project[]) {
  return function(fc: AbstractControl): any {
    const names = projects.map(p => p.name);
    return names.includes(fc.value) ? { uniqueProjectValidator: true } : null;
  };
}

@Component({
  selector: 'dcs-add-project',
  templateUrl: './add-project.component.html',
  styleUrls: ['./add-project.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class AddProjectComponent implements OnChanges {
  @Input()
  private projects: Project[];

  @Output()
  submitForm = new EventEmitter<{ name: string; description: string }>();

  @Output()
  cancel = new EventEmitter<any>();

  public form: FormGroup;

  constructor(fb: FormBuilder) {
    this.form = fb.group({
      name: ['', [Validators.required]],
      description: ['', [Validators.required]],
    });
  }

  public ngOnChanges(changes: SimpleChanges) {
    if (changes.projects) {
      this.form
        .get('name')
        .setValidators([Validators.required, uniqueProjectValidator(this.projects)]);
    }
  }

  public submit(): void {
    if (this.form.valid) {
      this.submitForm.next(this.form.value);
    }
  }
}
