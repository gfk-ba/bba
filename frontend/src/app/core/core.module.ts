import { registerLocaleData } from '@angular/common';
import { HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';
import localeDe from '@angular/common/locales/de';
import { InjectionToken, LOCALE_ID, NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { ServiceWorkerModule } from '@angular/service-worker';
import { APP_ENVIRONMENT } from '@dcs/ngx-tools';
import { StoreRouterConnectingModule } from '@ngrx/router-store';
import { ActionReducerMap, StoreModule } from '@ngrx/store';
import { StoreDevtoolsModule } from '@ngrx/store-devtools';
import { ApiInterceptor } from './api.interceptor';
import { environment } from '../../environments/environment';
import { metaReducers, reducers, State } from '../reducers';
import { LoginGuard } from '../reducers/login/login.guard';

export const REDUCER_TOKEN = new InjectionToken<ActionReducerMap<State>>('Registered Reducers');

export function getReducers() {
  return reducers;
}

@NgModule({
  imports: [
    BrowserModule,
    HttpClientModule,
    StoreModule.forRoot(REDUCER_TOKEN, { metaReducers }),
    !environment.production ? StoreDevtoolsModule.instrument() : [],
    StoreRouterConnectingModule.forRoot({
      stateKey: 'router',
    }),
    ServiceWorkerModule.register('/ngsw-worker.js', { enabled: environment.production }),
  ],
  declarations: [],
  providers: [
    { provide: LOCALE_ID, useValue: 'en' },
    { provide: APP_ENVIRONMENT, useValue: environment },
    {
      provide: REDUCER_TOKEN,
      useFactory: getReducers,
    },
    {
      provide: HTTP_INTERCEPTORS,
      useClass: ApiInterceptor,
      multi: true,
    },
    LoginGuard,
  ],
})
export class CoreModule {
  constructor() {
    registerLocaleData(localeDe);
  }
}
