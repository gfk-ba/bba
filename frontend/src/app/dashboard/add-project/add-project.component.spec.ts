import { Component, ViewChild } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { FormGroup, ReactiveFormsModule } from '@angular/forms';
import { AddProjectComponent } from './add-project.component';
import { configureComponentTestSuite } from '../../../testing';
import { Project } from '../../reducers/project/models/project.class';

@Component({
  selector: 'dcs-test-component',
  template: `
    <dcs-add-project
      [projects]="projects"
      (submitForm)="collectSubmit($event)"
    ></dcs-add-project>`,
})
class TestHostComponent {
  @ViewChild(AddProjectComponent)
  public subject: AddProjectComponent;

  public projects: Project[] = [];
  public submitData: any;

  public collectSubmit(data: any) {
    this.submitData = data;
  }
}

describe('AddProjectComponent', () => {
  let testHost: TestHostComponent;
  let fixture: ComponentFixture<TestHostComponent>;
  let subject: AddProjectComponent;
  let subjectEl: HTMLUnknownElement;

  configureComponentTestSuite();

  beforeAll(async () => {
    await TestBed.configureTestingModule({
      declarations: [TestHostComponent, AddProjectComponent],
      imports: [ReactiveFormsModule],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(TestHostComponent);
    testHost = fixture.componentInstance;
    subject = testHost.subject;
    subjectEl = fixture.nativeElement.querySelector('dcs-add-project');
    fixture.detectChanges();
  });

  it('should create testHost', () => {
    expect(testHost).toBeTruthy();
  });

  it('should render component', () => {
    expect(subjectEl).toBeInstanceOf(HTMLUnknownElement);
    expect(subject).toBeInstanceOf(AddProjectComponent);
  });

  describe('form', () => {
    it('creates the form', () => {
      expect(subject.form).toBeInstanceOf(FormGroup);
      expect(subject.form.value).toMatchSnapshot();
    });

    it('validates', () => {
      expect(subject.form.valid).toBeFalsy();
      subject.form.setValue({ name: 'testname', description: 'testdescription' });
      expect(subject.form.valid).toBeTruthy();
    });

    it('adds the uniqueProjectValidator', () => {
      testHost.projects = [new Project({ name: 'Project 1' }), new Project({ name: 'Project 2' })];
      fixture.detectChanges();

      subject.form.setValue({ name: 'Project 1', description: 'foo' });
      expect(subject.form.valid).toBeFalsy();

      subject.form.setValue({ name: 'Project 2', description: 'foo' });
      expect(subject.form.valid).toBeFalsy();

      subject.form.setValue({ name: 'Project 3', description: 'foo' });
      expect(subject.form.valid).toBeTruthy();
    });
  });

  describe('submitting the form', () => {
    describe('if invalid', () => {
      it('does not call the submitForm EventEmitter', () => {
        subject.form.setValue({ name: '', description: '' });
        subject.submit();

        expect(testHost.submitData).toBeUndefined();
      });
    });

    describe('if valid', () => {
      it('calls the submitForm EventEmitter', () => {
        subject.form.setValue({ name: 'testname', description: 'testdescription' });
        subject.submit();

        expect(testHost.submitData).toMatchSnapshot();
      });
    });
  });
});
