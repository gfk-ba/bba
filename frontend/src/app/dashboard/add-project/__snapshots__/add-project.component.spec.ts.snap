// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddProjectComponent form creates the form 1`] = `
Object {
  "description": "",
  "name": "",
}
`;

exports[`AddProjectComponent submitting the form if valid calls the submitForm EventEmitter 1`] = `
Object {
  "description": "testdescription",
  "name": "testname",
}
`;
